### PR TITLE
Pooled control message storage.

### PR DESCRIPTION
**Pull request by ser-0xff**
Pooled control message storage.

### Motivation:

Async I/O operations require related data remains until the operation is not completed.

### Modifications:

Control message data for UDP packets is pooled together with msghdr's.

### Result:

Async I/O can be used properly for UDP.

### CODE REVIEW

File: Sources/NIOPosix/SelectableEventLoop.swift [hunk 1]
```diff
@@ -185,7 +185,7 @@ Further information:
         self.thread = thread
         self.bufferPool = Pool<PooledBuffer>(maxSize: 16)
         self.msgBufferPool = Pool<PooledMsgBuffer>(maxSize: 16)
-        self.controlMessageStorage = UnsafeControlMessageStorage.allocate(msghdrCount: Socket.writevLimitIOVectors)
+        self.controlMessageStorage = UnsafeControlMessageStorage.allocate(msghdrCount: 1)
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/NIOPosix/SelectableEventLoop.swift:188]: There are few places where control message storage is used also. We will need to make changes like [that](https://github.com/apple/swift-nio/commit/f1edf0161105e5473e39c2770f0d47fa7ffccebc) to use pooled control message storage. That fix does not look well from my point of view, that is why I left a control message storage for one element in the SelectableEventLoop. If you are fine with those changes then I can add them to the PR.

[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIOPosix/SelectableEventLoop.swift:188]: Why do we have to do that? Why can't we borrow the pooled storage for the duration of the method?

[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/NIOPosix/SelectableEventLoop.swift:188]: Yes, we can.
As I showed in the [commit](https://github.com/apple/swift-nio/commit/f1edf0161105e5473e39c2770f0d47fa7ffccebc) borrowing pooled storage adds some complexity to the code. If you think that such complexity is ok, then I can add those changes and we will not have control message storage in the event loop. What do you think?

[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIOPosix/SelectableEventLoop.swift:188]: Sorry, I mean that this line is not acceptable: https://github.com/apple/swift-nio/commit/f1edf0161105e5473e39c2770f0d47fa7ffccebc#diff-a6ead8acc70ef0c19d2a1512f7d5642f51935a8621f2d68b10e3af24a84a184dR625. We aren't allowed to escape the pointer in this way. But I think we can just expand the scope of the closure to be much wider, can't we?

[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/NIOPosix/SelectableEventLoop.swift:188]: You are right.
What do you think about [that](https://github.com/ordo-one/swift-nio/commit/461d2eb534e2800b96d8cf9cfa1b42ac640f7cdc)?

[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/NIOPosix/SelectableEventLoop.swift:188]: Will add those changes to the PR then,
thank you!

**Comment by ser-0xff** [comment]
PR has been updated, please have a look.

File: Sources/NIOPosix/SelectableEventLoop.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/NIOPosix/SelectableEventLoop.swift]: PR has been updated, could you have a look?

File: Sources/NIOPosix/ControlMessage.swift [hunk 2]
```diff
@@ -25,31 +25,37 @@ import CNIOWindows
...
     let bytesPerMessage: Int
+    let deallocateBuffer: Bool
     var buffer: UnsafeMutableRawBufferPointer
 
     /// Initialise which includes allocating memory
     /// parameter:
     /// - bytesPerMessage: How many bytes have been allocated for each supported message.
     /// - buffer: The memory allocated to use for control messages.
-    private init(bytesPerMessage: Int, buffer: UnsafeMutableRawBufferPointer) {
+    init(bytesPerMessage: Int, deallocateBuffer: Bool, buffer: UnsafeMutableRawBufferPointer) {
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/NIOPosix/ControlMessage.swift]: Fixed.

File: Sources/NIOPosix/ControlMessage.swift [hunk 3]
```diff
@@ -25,31 +25,37 @@ import CNIOWindows
 /// Supports multiple messages each with enough storage for multiple `cmsghdr`
 struct UnsafeControlMessageStorage: Collection {
     let bytesPerMessage: Int
+    let deallocateBuffer: Bool
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/NIOPosix/ControlMessage.swift]: Fixed.

File: Sources/NIOPosix/ControlMessage.swift [hunk 4]
```diff
@@ -25,31 +25,37 @@ import CNIOWindows
...
         let buffer = UnsafeMutableRawBufferPointer.allocate(byteCount: bytesPerMessage * msghdrCount,
                                                              alignment: MemoryLayout<cmsghdr>.alignment)
-        return UnsafeControlMessageStorage(bytesPerMessage: bytesPerMessage, buffer: buffer)
+        return UnsafeControlMessageStorage(bytesPerMessage: bytesPerMessage, deallocateBuffer: true, buffer: buffer)
     }
 
     mutating func deallocate() {
-        self.buffer.deallocate()
-        self.buffer = UnsafeMutableRawBufferPointer(start: UnsafeMutableRawPointer(bitPattern: 0x7eadbeef), count: 0)
+        if self.deallocateBuffer {
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/NIOPosix/ControlMessage.swift:64]: Let me try to do it in the next PR.

**Comment by ser-0xff** [comment]
PR has been updated, please have a look.

File: Sources/NIOPosix/ControlMessage.swift [hunk 5]
```diff
@@ -26,30 +26,45 @@ import CNIOWindows
...
+                                                            alignment: MemoryLayout<cmsghdr>.alignment)
+        return UnsafeControlMessageStorage(bytesPerMessage: bytesPerMessage, buffer: buffer, deallocateBuffer: true)
+    }
+
+    /// Create an instance not owning the buffer
+    /// parameter:
+    /// - bytesPerMessage: How many bytes have been allocated for each supported message.
+    /// - buffer: The memory allocated to use for control messages.
+    static func makeNotOwning(bytesPerMessage: Int, buffer: UnsafeMutableRawBufferPointer) -> UnsafeControlMessageStorage {
+        return UnsafeControlMessageStorage(bytesPerMessage: bytesPerMessage, buffer: buffer, deallocateBuffer: false)
```
Comment (Lukasa) [inline Sources/NIOPosix/ControlMessage.swift:60]: Here can we add a `precondition` that `buffer.count` is large enough?

Comment (ser-0xff) [reply Sources/NIOPosix/ControlMessage.swift:61]: Added precondition for buffer size, unfortunately at that point it is not known for how many messages user supposed to use that control message storage, so the only we can check is buffer size is not less than bytesPerMessage.

**Pull request merged**